### PR TITLE
Cross Browser Gradient for App Drawer

### PR DESCRIPTION
**Pull request by TheLoneRonin**
### Description

- Uses `backdrop-filter` when the Browser can support it.

- Removes the `.png` image for a CSS gradient.

- Adds the shine effect as seen on: https://wiki.metagame.wtf/docs/home when you hover over item elements.

#### Chrome

https://user-images.githubusercontent.com/68780068/105202099-9d804b80-5b0f-11eb-9a8a-e7708846f580.mp4

#### Firefox

https://user-images.githubusercontent.com/68780068/105202111-a244ff80-5b0f-11eb-9865-9e5ed788aa0a.mp4

-- EDIT

paired with @dysbulic

### CODE REVIEW

**Comment by TheLoneRonin** [comment]
Before merging, need to make sure that the preview is the same as the dev environment.
For some weird reason the preview does not have filter gradient on Chrome.

Just want to confirm is this potentially my cache? Hopefully not!

**Comment by dysbulic** [comment]
Chromium is still borked:

Screenshot from 2021-01-21 03-37-00

Chrome looks good though under Linux:

Screenshot from 2021-01-21 03-38-02

File: packages/web/components/PageHeader.tsx [hunk 1]
```diff
@@ -217,9 +243,30 @@ export const PageHeader: React.FC = () => {
...
+            }
+            style={{ backdropFilter: 'blur(10px)' }}
+          />
+          <Image
+            position="absolute"
+            left="calc(50% - 16.5rem)"
+            top={678}
+            src={MetaGradientBottom}
+            alt="MetaGradientBottom"
+            width={522}
```
Comment (dysbulic) [inline packages/web/components/PageHeader.tsx:266]: Should `left` and `width` be in the same units so there's no possibility of misaligning things.

(I'm assuming you're centering something.)

**Review by dysbulic — APPROVED** [review]

File: packages/web/components/PageHeader.tsx [hunk 2]
```diff
@@ -217,9 +243,30 @@ export const PageHeader: React.FC = () => {
...
-          <Box position="absolute" left="calc(50% - 16.5rem)" top="0">
-            <NextImage src={MetaBox} alt="MetaBox" width={528} height={695} />
-          </Box>
+
+          <Box
+            position="absolute"
+            left="calc(50% - 16.5rem)"
+            top="0"
+            width={522}
+            height={680}
```
Comment (hammadj) [inline packages/web/components/PageHeader.tsx:252]: Can we use responsive sizing instead of hard coded width/heights? Hardcoded sizes is why the drawer is breaking / looking weird on a lot of devices

File: packages/web/components/PageHeader.tsx (hunk 1 above)
Comment (hammadj) [inline packages/web/components/PageHeader.tsx:266]: Shouldn't use hardcoded values. Also, Chakra has a native Gradient component now, shouldn't use an image since it takes a lot more bandwidth to load vs plain CSS.

File: packages/web/components/PageHeader.tsx [hunk 3]
```diff
@@ -77,11 +79,42 @@ const SubMenuItem: React.FC<React.ComponentProps<typeof MetaLink>> = ({
...
+          position: 'absolute',
+          display: 'block',
+          top: 0,
+          left: 0,
+          width: '100%',
+          height: '100%',
+          filter: 'blur(10px)',
+          transform: 'translate3d(-120%, 0, 0)',
+          background:
+            'linear-gradient(45deg, transparent, rgba(255, 255, 255, 0), rgba(255, 255, 255, 0.06), rgba(255, 255, 255, 0), transparent)',
```
Comment (hammadj) [inline packages/web/components/PageHeader.tsx:109]: A lot of repeated code / styles, should consolidate and clean it up. You dont need 90% of these if you follow Chakra UI method of styling

**Review by hammadj — CHANGES_REQUESTED** [review]

**Comment by dysbulic** [comment]
@TheLoneRonin, @alalonde asked if I would [look at the display issues with Chromium](https://github.com/MetaFam/TheGame/issues/254#issuecomment-766457797). I was going to work off this branch since @hammadj suggested it is [the use of absolute units that is the root of the problem](https://github.com/MetaFam/TheGame/pull/294#discussion_r562357561).

The tasks are:

- [ ] leverage Chakra UI to reduce the amount of repeated styling
- [ ] change absolute units to relative ones
- [ ] use the native Chakra UI gradient component

It's entirely possible you've already done this stuff. If so, just drop me a line.

File: packages/web/components/PageHeader.tsx (hunk 2 above)
Comment (dysbulic) [reply packages/web/components/PageHeader.tsx]: Things are now in terms of `rem`s. String interpolation is used along with `calc()` to create comprehensible sizing instructions.

It's feasible that some more of those statements could be derived from `flex` or `grid` layouts, but the current implementation works.

File: packages/web/components/PageHeader.tsx (hunk 1 above)
Comment (dysbulic) [reply packages/web/components/PageHeader.tsx]: The image is the bottom cap on the drawer. The background is a CSS gradient.

File: packages/web/components/PageHeader.tsx (hunk 3 above)
Comment (dysbulic) [reply packages/web/components/PageHeader.tsx]: The bulk of the repetition was a `_before` with a `display` of `none`. Removing it didn't affect the display, so I removed it.

**Comment by pakokrew** [comment]
- Drop blur doesn't work on my browser (Brave 1.19) and even making weird flickering.

- Button logos are still starting to load with a wrong placeholder image. That was fixed by replacing `Image` component by [the one of nextjs ](https://nextjs.org/docs/api-reference/next/image), but it look like what I've done have been reverted.


Uploading Screen Recording 2021-01-31 at 13.22.41.mov…


Uploading Screen Recording 2021-01-31 at 13.21.07.mov…

**Comment by dysbulic** [comment]
> - Button logos are still starting to load with a wrong placeholder image. That was fixed by replacing Image component by the one of nextjs , but it look like what I've done have been reverted.

The issue with the Next.js `Image` is that the dimensions have to be specified in pixels and switching to a responsive layout meant that they weren't.

I can hopefully simply put a sized `div` around it and then use `Image`'s `layout='fill'`.

**Review by hammadj — CHANGES_REQUESTED** [review]
Blur not working in Safari, CI failing due to linting errors

image

File: packages/web/components/MobileFooter.tsx [hunk 4]
```diff
@@ -93,66 +89,65 @@ export const MobileFooter: React.FC = () => {
...
+            display="flex"
+            textDecoration="none"
+            variant="link"
+            flexDirection="column"
+            justifyContent="center"
+            alignItems="center"
+            onClick={onToggle}
+            _focus={{ boxShadow: 'none' }}
+          >
+            <Image // TODO use NextImage component once images are without text
```
Comment (pakokrew) [inline packages/web/components/MobileFooter.tsx:134]: Use next/image

File: packages/web/components/PageHeader.tsx [hunk 5]
```diff
@@ -90,6 +116,27 @@ const SubMenuItem: React.FC<React.ComponentProps<typeof MetaLink>> = ({
   );
 };
 
+const BoxedNextImage: React.FC<
```
Comment (pakokrew) [inline packages/web/components/PageHeader.tsx:119]: Don't know if this is good practice, but that's a good workaround for next/image !

Maybe we could put this in the `design-system` package instead ? Or at least in a component of its own in `web`, so it could be reusable (🔔 MobileFooter)

File: packages/web/components/MobileFooter.tsx [hunk 6]
```diff
@@ -93,66 +89,65 @@ export const MobileFooter: React.FC = () => {
...
-          variant="link"
-          flexDirection="column"
-          justifyContent="center"
-          alignItems="center"
-          width="20vw"
-          height="5rem"
-          onClick={onToggle}
+        {DrawerItemsLeft.map((item) => (
+          <MenuItem href={item.href} isExternal={item.isExternal} key={item.href}>
+            <NextImage src={item.src} alt={item.alt} width={35} height={35} />
```
Comment (pakokrew) [inline packages/web/components/MobileFooter.tsx:107]: these are still with fixed size maybe use new `BoxedNextImage` component

File: packages/web/components/MobileFooter.tsx [hunk 7]
```diff
@@ -93,66 +89,65 @@ export const MobileFooter: React.FC = () => {
...
 
-      {DrawerItemsRight.map((item) => (
-        <MenuItem href={item.href} isExternal={item.isExternal} key={item.href}>
-          <NextImage src={item.src} alt={item.alt} width={35} height={35} />
-          {item.text}
-        </MenuItem>
-      ))}
+        {DrawerItemsRight.map((item) => (
+          <MenuItem href={item.href} isExternal={item.isExternal} key={item.href}>
+            <NextImage src={item.src} alt={item.alt} width={35} height={35} />
```
Comment (pakokrew) [inline packages/web/components/MobileFooter.tsx:146]: same as l.107

File: packages/web/components/MobileFooter.tsx [hunk 8]
```diff
@@ -162,28 +157,30 @@ export const MobileFooter: React.FC = () => {
...
-            return (
-              <SubMenuItem href={item.href} key={item.alt}>
-                <SubImage src={item.src} alt={item.alt} />
-                {item.text}
-              </SubMenuItem>
-            );
-          })}
+          {DrawerSubItems.map((item) => (
+            <SubMenuItem href={item.href} key={item.alt}>
+              <SubImage src={item.src} alt={item.alt} />
```
Comment (pakokrew) [inline packages/web/components/MobileFooter.tsx:180]: BoxedNextImage

File: packages/web/components/PageHeader.tsx [hunk 9]
```diff
@@ -217,17 +272,46 @@ export const PageHeader: React.FC = () => {
...
+                `linear-gradient(
+                  180deg, rgba(76, 63, 143, ${opacity}) 62.76%,
+                  rgba(184, 169, 255, ${opacity}) 100%
+                )`
+              )
+            })()}
+            style={{ backdropFilter: 'blur(10px)' }}
+          />
+          {/* ToDo: switch to BoxedNextImage */}
+          <Image
```
Comment (pakokrew) [inline packages/web/components/PageHeader.tsx:294]: BoxedNextImage

File: packages/web/components/PageHeader.tsx [hunk 10]
```diff
@@ -100,37 +147,36 @@ export const PageHeader: React.FC = () => {
...
-        <Box position="absolute" left="calc(50% - 23.5rem)" top="0" zIndex="1">
-          <NextImage
+        {/* margin-inline-start = 0.5rem is set and I don't know how to remove it */}
+        <Box
+          position="absolute"
+          left={`calc(50% - (${drawer.width} / 2) + 0.5rem)`}
+          top="0"
+          zIndex="1"
+        >
+          <Image
```
Comment (pakokrew) [inline packages/web/components/PageHeader.tsx:175]: BoxedNextImage

**Review by pakokrew — CHANGES_REQUESTED** [review]
Few images components were forgotten.
Also some linting errors.

On brave 1.19, drop blur filter works for in MobileFooter but not in PageHeader

Good luck, we're almost good !

File: packages/web/components/PageHeader.tsx (hunk 5 above)
Comment (dysbulic) [reply packages/web/components/PageHeader.tsx:119]: My understanding is that `NextImage`'s primary function is to load the lowest resolution image possible for a given display size.

It is using [`scrset` and `sizes`](https://developer.mozilla.org/en-US/docs/Learn/HTML/Multimedia_and_embedding/Responsive_images#resolution_switching_different_sizes), but since it doesn't enlarge an image past its original resolution that isn't really affecting these images because they're so small.

I'll:
- [ ] move the `BoxedNextImage` to the design system
- [ ] change the references in the page header
- [ ] add usage in the page footer

**Comment by dysbulic** [comment]
> Blur not working in Safari, CI failing due to linting errors

I'm pretty sure the problem is that Next.js is rendering the pages server-side where `background-filter` isn't supported. This disables the opacity of the drawer.

I [posted to StackOverflow](https://stackoverflow.com/questions/66017821/how-to-handle-unsupported-css-properties-in-next-js-or-use-supports-directives) on the issue.

File: packages/web/components/PageHeader.tsx (hunk 5 above)
Comment (pakokrew) [reply packages/web/components/PageHeader.tsx]: I didn't dig into the details of the next/image component, and maybe I'm wrong saying it's better to use it for that particular case.

What is sure, is that there shouldn't be a placeholder image like we see in the video I uploaded while they are loading. This is not the case in other parts of the app. So creating a custom image component is a good idea, where we will be able to switch either to next/image or to a custom solution solving that issue.

File: packages/web/components/LoginButton.tsx [hunk 11]
```diff
@@ -45,16 +46,25 @@ export const LoginButton: React.FC = () => {
             {user.player ? getPlayerName(user.player) : 'Unknown'}
           </MetaLink>
           <HStack spacing={2}>
-            <MetaLink href="/profile/setup">Setup profile</MetaLink>
+            <MetaLink
```
Comment (pakokrew) [inline packages/web/components/LoginButton.tsx:49]: Why these changes in this branch ?

**Comment by pakokrew** [comment]
I think I solved the issue. There was an error in the web console:
`next js Prop `className` did not match. Server: css-azed Client:css-12J21`

Problem is well explain here: https://www.joshwcomeau.com/react/the-perils-of-rehydration/
The page rendered by the server and the first render of react MUST be the same, otherwise react is lost and can't rehydrate correctly.

File: packages/web/components/LoginButton.tsx (hunk 11 above)
Comment (dysbulic) [reply packages/web/components/LoginButton.tsx:49]: When you narrow the screen when the profile options are written out they are wide enough they push the profile picture behind the drawer. The switch to icons was to narrow everything.

**Comment by hammadj** [comment]
> I think I solved the issue. There was an error in the web console:
> `next js Prop 'className' did not match. Server: css-azed Client:css-12J21`
> 
> Problem is well explain here: https://www.joshwcomeau.com/react/the-perils-of-rehydration/
> The page rendered by the server and the first render of react MUST be the same, otherwise react is lost and can't rehydrate correctly.
> 
> It is left to replace all the images component that doesn't use the new one.

Works in Chrome but still doesn't work in Safari. Here's the code I usually use for background blur that works in Safari as well: https://github.com/expo/expo/blob/sdk-40/packages/expo-blur/src/BlurView.web.tsx. We could literally just import that expo/blur library and use it instead of doing all our own implementation too.

Screenshot from Safari of that BlurView component: 
image

Screenshot of AppDrawer in Safari:
image

**Comment by dysbulic** [comment]
@hammadj, Safari needs the `-webkit-` prefix for `backdropFilter`. I added it, so it should be sorted.

File: packages/web/components/MobileFooter.tsx (hunk 4 above)
Comment (dysbulic) [reply packages/web/components/MobileFooter.tsx]: Done.

File: packages/web/components/MobileFooter.tsx (hunk 6 above)
Comment (dysbulic) [reply packages/web/components/MobileFooter.tsx]: Done.

File: packages/web/components/MobileFooter.tsx (hunk 7 above)
Comment (dysbulic) [reply packages/web/components/MobileFooter.tsx]: Done.

File: packages/web/components/PageHeader.tsx (hunk 9 above)
Comment (dysbulic) [reply packages/web/components/PageHeader.tsx:285]: This one is the only one that's tricky. It has `max-width`  and `max-height` set to `100%`. It will expand beyond it's original dimensions on the y-axis, but not on the x.

I went into the debugger and added random CSS props & tried lots of different values, but it just wouldn't line up when using `BoxedNextImage`.

File: packages/web/components/PageHeader.tsx (hunk 10 above)
Comment (dysbulic) [reply packages/web/components/PageHeader.tsx:163]: Done.

File: packages/web/components/MobileFooter.tsx (hunk 8 above)
Comment (dysbulic) [reply packages/web/components/MobileFooter.tsx:221]: Done.

File: packages/web/pages/guild/[guildname].tsx [hunk 12]
```diff
@@ -81,7 +81,7 @@ export const getStaticProps = async (
 
   return {
     props: {
-      guild,
+      guild: guild || null,
```
Comment (dysbulic) [inline packages/web/pages/guild/[guildname].tsx]: This will give the incorrect output when `guid` is `false` (or anything other than undefined).

**Review by dysbulic — APPROVED** [review]
I'm a little bleary eyed, but I see nothing glaringly wrong with this code.

I've tested a reasonable amount on the browsers available to me.

**Comment by alalonde** [comment]
Just checked the latest deployment, backdrop blur effect looks fine on safari/FF mac. lgtm

**Review by pakokrew — APPROVED** [review]

**Comment by dysbulic** [comment]
@TheLoneRonin, I'd like to split the credit with this with you. ¿Is that acceptable?

@other-author: @dysbulic

**Comment by pakokrew** [comment]
> @TheLoneRonin, I'd like to [split the credit](https://sourcecred.io/docs/beta/plugins/github/#edges:~:text=%40other%2Dauthor) with this with you. ¿Is that acceptable?
> 
> @other-author: @dysbulic

I've already allowed myself to edit the main post to credit you 😇

**Comment by pakokrew** [comment]
Let's merge this.

**Comment by peth-yursick** [comment]
> 
> 
> Let's merge this.

still shows gradient on Firefox on Windows
